spec/requests/halls_spec.rb

```diff
@@ -0,0 +1,345 @@
+      before { sign_in user }
+      it "redirects to root" do
+        get("/halls/#{hall.id}/edit")
+        expect(response).to redirect_to("/")
+      end
+
+      it "returns redirect status" do
+        get("/halls/#{hall.id}/edit")
+        expect(response.status).to eq(302)
+      end
+    end
+
+    context "when user with permission" do
+      before { sign_in manager }
+      it "returns successful response" do
+        get("/halls/#{hall.id}/edit")
+        expect(response.status).to eq(200)
+      end
+
+      it "renders proper template" do
+        get("/halls/#{hall.id}/edit")
+        expect(response).to render_template("halls/edit")
+      end
+    end
+  end
+
+  describe "PATCH /halls" do
+    let(:hall) { create :hall }
+    let(:params) do
+      {
+        hall: {
+          name:,
+          capacity:
+        }
+      }
+    end
+
+    let(:name) { "Sala #{Faker::Number.number(digits: 1)}" }
+    let(:capacity) { Faker::Number.number(digits: 2) }
+
+    context "when no user" do
+      it "redirects to sign_in" do
+        patch("/halls/#{hall.id}", params:)
+        expect(response).to redirect_to("/users/sign_in")
+      end
+
+      it "returns redirect status" do
+        patch("/halls/#{hall.id}")
+        expect(response.status).to eq(302)
+      end
+    end
+
+    context "when user without permission" do
+      before { sign_in user }
+      it "redirects to root" do
+        patch("/halls/#{hall.id}", params:)
+        expect(response).to redirect_to("/")
+      end
+
+      it "returns redirect status" do
+        patch("/halls/#{hall.id}")
+        expect(response.status).to eq(302)
+      end
+    end
+
+    context "when user with permission" do
+      before { sign_in manager }
+      it "returns redirect status" do
+        patch("/halls/#{hall.id}", params:)
+        expect(response.status).to eq(302)
+      end
+
+      it "redirects to /halls" do
+        patch("/halls/#{hall.id}", params:)
+        expect(response).to redirect_to("/halls")
+      end
+
+      it "updates hall record" do
+        expect { patch("/halls/#{hall.id}", params:) }.to change { hall.reload.name }.from(hall.name).to(name.to_s)
+      end
+    end
+
+    context "when params invalid" do
+      before { sign_in manager }
+      let(:duration) { "s" }
+
+      it "doesn't create hall record" do
+        expect { patch("/halls/#{hall.id}", params:) }.not_to(change { hall.reload.name })
+      end
+
+      it "returns unsuccessful response" do
+        patch("/halls/#{hall.id}", params:)
+        expect(response.status).to eq(422)
+      end
+
+      let(:duration) { Faker::Number.number(digits: 2) }
+      let(:name) { nil }
+
+      it "doesn't create hall record" do
+        expect { patch("/halls/#{hall.id}", params:) }.not_to(change { hall.reload.name })
+      end
+
+      it "returns unsuccessful response" do
+        patch("/halls/#{hall.id}", params:)
+        expect(response.status).to eq(422)
+      end
+    end
+  end
```

Comment:
I would split this into two contexts
`when duration too short` and `when name nil`

Reply:
Yup, looks clearer, i don't change context in the middle and don't redefine variable now. Much better :).